plugin/compiler-plugin/src/main/kotlin/it/unibo/collektive/FieldProjectionVisitor.kt

```diff
@@ -0,0 +1,72 @@
+package it.unibo.collektive
+
+import it.unibo.collektive.utils.common.AggregateFunctionNames.FIELD_CLASS
+import it.unibo.collektive.utils.common.AggregateFunctionNames.PROJECT_FUNCTION
+import it.unibo.collektive.utils.common.putTypeArgument
+import it.unibo.collektive.utils.logging.error
+import it.unibo.collektive.utils.statement.irStatement
+import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
+import org.jetbrains.kotlin.cli.common.messages.MessageCollector
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrFunction
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.IrGetValue
+import org.jetbrains.kotlin.ir.expressions.putArgument
+import org.jetbrains.kotlin.ir.types.classFqName
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+
+class FieldProjectionVisitor(
+    private val pluginContext: IrPluginContext,
+    private val logger: MessageCollector,
+    private val aggregateContextClass: IrClass,
+    private val aggregateReference: IrExpression?,
+) : IrElementTransformerVoid() {
+
+    private val projectFunction by lazy {
+        aggregateContextClass.functions
+            .filter { it.name == Name.identifier(PROJECT_FUNCTION) }
+            .firstOrNull()
+    }
+
+    override fun visitGetValue(expression: IrGetValue): IrExpression {
+        if (expression.type.classFqName == FqName(FIELD_CLASS)) {
+            aggregateReference?.let { aggregateRef ->
+                projectFunction?.let { projectFunc ->
+                    return wrapInProjectFunction(expression, aggregateRef, projectFunc)
+                } ?: logger.error(
+                    """
+                        Fail to look up the `project` function to perform the field projection.
+                        This can happen if the `AggregateContext` class is not found by the compiler plugin.
+                    """.trimIndent()
+                )
+            } ?: logger.error(
+                """
+                    Fail to look up the aggregate context reference to perform the field projection.
+                    This can happen if the aggregate context reference is not available in the current scope,
+                    and likely means that a field operation is being performed outside of an aggregate scope.
+                """.trimIndent()
```

Comment:
Can this happen? Shouldn't the type checker prevent it?

Reply:
In practice should be never thrown. However the look up process in the compiler plugin may fail and I have managed the error in that way.

Reply:
Okay, makes sense. I was just wandering who could you possibly had already walked into it